presto-spi/src/main/java/com/facebook/presto/spi/type/SqlTimeWithTimeZone.java

```diff
@@ -119,18 +113,36 @@ public boolean equals(Object obj)
     @Override
     public String toString()
     {
+        Instant utcInstantInCurrentDay = getTimeInstantInCurrentDay();
+        ZonedDateTime zonedDateTime = utcInstantInCurrentDay.atZone(ZoneId.of(timeZoneKey.getId()));
+        return zonedDateTime.format(formatter);
+    }
+
+    private Instant getTimeInstantInCurrentDay()
+    {
         // FIXME This is hack that we need to use as the timezone interpretation depends on date (not only on time)
         // Additionally we cannot just grab current TZ offset from timezoneOffsetReferenceTimestampUtc and use it
         // as our formatting library will fail to display expected TZ symbol.
         long currentMillisOfDay =
-                LocalTime.from(
+                TimeUnit.NANOSECONDS.toMillis(LocalTime.from(
                         Instant.ofEpochMilli(timezoneOffsetReferenceTimestampUtc)
                                .atZone(ZoneOffset.UTC))
-                        .toNanoOfDay() / 1_000_000L; // nanos to millis
+                        .toNanoOfDay());
         long timeMillisUtcInCurrentDay = timezoneOffsetReferenceTimestampUtc - currentMillisOfDay + millisUtc;
+        return Instant.ofEpochMilli(timeMillisUtcInCurrentDay);
+    }
 
-        Instant utcInstantInCurrentDay = Instant.ofEpochMilli(timeMillisUtcInCurrentDay);
-        ZonedDateTime zonedDateTime = utcInstantInCurrentDay.atZone(ZoneId.of(timeZoneKey.getId()));
-        return zonedDateTime.format(formatter);
+    private void validateMillisRange()
+    {
+        if (getLocalMillis() < MIN_LOCAL_MILLIS || getLocalMillis() >= MAX_LOCAL_MILLIS) {
+            throw new RuntimeException("millisUtc + TZ offset must be between 0:00-23:59:59.999 UTC (" + this.millisUtc + " in time zone " + this.timeZoneKey + "is to high)");
+        }
+    }
+
+    private long getLocalMillis()
+    {
+        ZoneId id = ZoneId.of(timeZoneKey.getId());
+        ZoneOffset offset = id.getRules().getOffset(getTimeInstantInCurrentDay());
+        return millisUtc + TimeUnit.SECONDS.toMillis(offset.getTotalSeconds());
```

Comment:
This is why they both use getTimeInstantInCurrentDay(), the implementation is forked afterwards.